pkg/controller/incompatible_operator_controller.go

```diff
@@ -0,0 +1,181 @@
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	storage "github.com/operator-framework/helm-operator-plugins/pkg/storage"
+	"github.com/operator-framework/operator-registry/alpha/property"
+	helm "helm.sh/helm/v3/pkg/storage"
+	"helm.sh/helm/v3/pkg/storage/driver"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+	v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/klog/v2"
+)
+
+const (
+	reasonIncompatibleOperatorsInstalled = "IncompatibleOperatorsInstalled"
+	typeIncompatibelOperatorsUpgradeable = "InstalledOLMOperatorsUpgradeable"
+	reasonInvalidBundleProperties        = "InvalidBundleProperties"
+	maxOpenShiftVersionProperty          = "olm.maxOpenShiftVersion"
+	ownerKindKey                         = "olm.operatorframework.io/owner-kind"
+	ownerNameKey                         = "olm.operatorframework.io/owner-name"
+	packageNameKey                       = "olm.operatorframework.io/package-name"
+	bundleNameKey                        = "olm.operatorframework.io/bundle-name"
+	bundleVersionKey                     = "olm.operatorframework.io/bundle-version"
+)
+
+type incompatibleOperatorController struct {
+	name                   string
+	nextOCPMinorVersion    *semver.Version
+	kubeclient             kubernetes.Interface
+	clusterExtensionClient *clients.ClusterExtensionClient
+	operatorClient         *clients.OperatorClient
+	logger                 logr.Logger
+}
+
+func NewIncompatibleOperatorController(name string, nextOCPMinorVersion *semver.Version, kubeclient kubernetes.Interface, clusterExtensionClient *clients.ClusterExtensionClient, operatorClient *clients.OperatorClient, eventRecorder events.Recorder) factory.Controller {
+	c := &incompatibleOperatorController{
+		name:                   name,
+		nextOCPMinorVersion:    nextOCPMinorVersion,
+		kubeclient:             kubeclient,
+		clusterExtensionClient: clusterExtensionClient,
+		operatorClient:         operatorClient,
+		logger:                 klog.NewKlogr().WithName(name),
+	}
+
+	return factory.New().WithSync(c.sync).WithSyncDegradedOnError(operatorClient).WithInformers(operatorClient.Informer(), clusterExtensionClient.Informer().Informer()).ToController(name, eventRecorder)
+}
+
+func (c *incompatibleOperatorController) sync(ctx context.Context, _ factory.SyncContext) error {
+	c.logger.Info("sync started")
+	defer c.logger.Info("sync finished")
+
+	var updateStatusFn v1helpers.UpdateStatusFunc
+	incompatibleOperators, err := c.getIncompatibleOperators()
+	if len(incompatibleOperators) > 0 {
+		updateStatusFn = v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    typeIncompatibelOperatorsUpgradeable,
+			Status:  operatorv1.ConditionFalse,
+			Reason:  reasonIncompatibleOperatorsInstalled,
+			Message: strings.Join(incompatibleOperators, ","),
```

Comment:
Provide more context in this message? I know in the past users have been confused and think that patch versions are blocked because of this. 
```
fmt.Sprintf("Found ClusterExtensions that require upgrades prior to upgrading cluster to version %d.%d: %s", nextOCPVersion.Major, nextOCPVersion.Minor, strings.Join(incompatibleOperators, ","))
```

Reply:
Also, I think that if `err != nil` here, we should still include it's message to give users that context as early as possible. Like "Your cluster has incompatible operators installed, so Upgradeable=False.... but also you've got these other problems that will need your attention too"